### PR TITLE
Rust Once, Run Everywhere

### PR DESCRIPTION
**Pull request by alexcrichton**
(don't merge until tomorrow!)

### CODE REVIEW

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+subtle aspects in play. First off we've got our function definition with a
+`#[no_mangle]` attribute. This instructs the compiler to not mangle the symbol
+name for the function `double_input`. Rust employs name mangling similar to C++
+to ensure that libraries do not clash with one another, and this attributes
+means that you don't have to guess a symbol name like
+`double_input::h485dee7f568bebafeaa` from C.
+
+Next we've got our function definition, and the most interesting part about
+this is the keyword `extern`. This is a specialized form of specifying the [ABI
+for a function][abi-fn] which enables C to know how to call this function.
```
Comment (sfackler) [inline _posts/2015-04-24-FFI-and-Rust.md]: We're not really teaching C how to call it, but building it in a form that C knows how to call. Maybe rephrase it to "which causes the function to work in the way C expects"?

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 2]
```diff
@@ -0,0 +1,264 @@
+---
+layout: post
+title: "FFI and Rust"
+author: Alex Crichton
+description: "Zero-cost and safe FFI in Rust"
+---
+
+
+Rust's quest for world domination was never destined to happen overnight, so
```
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: Yes. World domination.

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+Rust needs to be able to interoperate with the the existing world just as easily
+as it talks to itself. To solve this problem, **Rust lets you communicate with C
+APIs at no extra cost while providing strong safety guarantees**.
+
+This is also referred to as Rust's foreign function interface (FFI) and is the
+method by which Rust communicates with other programming languages. In following
+Rust's design principles, this is a **zero cost abstraction** which leverages
+language features such as ownership and borrowing to provide a **safe
+interface**.
```
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: I don't see how the FFI itself leverages ownership and borrowing.

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+In this post we'll explore how Rust's zero-cost FFI and ability to write safe
+abstractions by looking at some examples of interacting with C. C is, however,
```
Comment (huonw) [inline _posts/2015-04-24-FFI-and-Rust.md]: "C. C" looks a little strange.

**Comment by huonw** [comment]
I'd still like to see at least a mention of destructors in terms of building good abstractions around FFI.

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+just an example, as we'll also see how Rust can easily talk to languages like
+Python and Ruby just as seamlessly as C.
```
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: First sentence includes a stray 'how' that is never connected to another clause. Maybe "In this post we'll explore how to encapsulate unsafe FFI calls to C in *safe, zero-cost abstractions*."

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+### Talking to C
+
+First, let's start with an example of calling C code from Rust and then take a
+look at how there is no overhead. Starting off simple, here's a C program which
```
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: Instead of 'take a look at how' you might say 'demonstrate that'. Instead of just 'no overhead', which is somewhat ambiguous, maybe 'Rust imposes no additional performance overhead', or 'the machine code generated by Rust to call C is equivalent to that generated by a C compiler'.

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    let input = 4;
+    let output = unsafe { double_input(input) };
+    println!("{} * 2 = {}", input, output);
+}
+```
+
+And that's it! You can try this out for yourself by [checking out the code on
+GitHub][rust2c] and running `cargo run` from that directory. As you can see
+there is 0 setup necessary to call into a C function and there's no overhead on
+Rust's side when calling it. There are, however, a few subtle aspects of this
```
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: I don't know that I 'see' there's no overhead. The compiler could be doing anything.

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+fn main() {
+    let input = 4;
+    let output = unsafe { double_input(input) };
+    println!("{} * 2 = {}", input, output);
+}
+```
+
+We see one of the crucial aspects of FFI in Rust here, the `unsafe` block. The
+compiler knows nothing about the implementation of `double_input`, so it must
+assume that memory unsafety *could* happen in this scenario. This may seem
```
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: Instead of 'memory unsafety could happen' you might embellish a colorful scenario like 'so it must assume that `double_input` may vomit nasal-demons all over your heap, which would clearly be memory-unsafe'.

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+limiting, but Rust has just the right set of tools to allow consumers to not
+worry about `unsafe` (more on this in a moment).
+
+Now that we've seen how to call a C function from Rust, let's see if we can
+verify this claim of zero overhead. Almost all programming languages can call
+into C one way or another, but it often comes at a cost with runtime type
+coversions or perhaps some language runtime juggling. To get a handle on what
```
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: 'coversions'

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+Rust is doing, let's go straight to the assembly code of the above `main`
+function:
+
+```
+mov    $0x4,%edi
+callq  3bc30 <double_input>
+```
```
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: This isn't the complete 'assembly' of `main`. May be worth clarification.

File: _posts/2015-04-24-FFI-and-Rust.md (hunk 3 above)
Comment (alexcrichton) [inline _posts/2015-04-24-FFI-and-Rust.md]: Hm I was trying to get both points up here to start emphasizing them both right off the bat, but I do agree that the wording here is a little odd. I'll see what I can do about this.

File: _posts/2015-04-24-FFI-and-Rust.md (hunk 8 above)
Comment (alexcrichton) [inline _posts/2015-04-24-FFI-and-Rust.md]: Hm I feel like this would need a different example, however. For example if we were passing in a pointer then C holding onto it or erroneously modifying it would be easy and bad to do, but when passing in an integer there's only so many bad things can happen. I do agree though it would be nice to emphasize this a bit more.

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+method returns a **borrowed slice** whose lifetime is connected to the same
+lifetime as the source tarball itself. This is Rust's way of indicating that the
+returned data cannot outlive the tarball, statically preventing bugs that may be
+encountered when just using C.
+
+A key aspect of the Rust binding here is that it is a safe function! Although it
+has an `unsafe` implementation (due to calling an FFI function), this interface
+is safe to call and will not cause tough-to-track-down segfaults. And don't
+forget, all of this is coming at 0 cost as the raw types in C are representable
+in Rust with no extra allocations or overhead.
```
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: This safe interface imposes an extra branch over using the raw interface because it checks for a null return and the caller is going to have to duplicate that check of the Option.

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+
+### Talking to Rust
+
+A major feature of Rust is that it does not have a garbage collector or heavy
+runtime, and one of the benefits of this is that Rust can be called from C with
```
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: May not want to equivocate here and just say 'does not have a garbage collector or runtime'.

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+
+Next we've got our function definition, and the most interesting part about
+this is the keyword `extern`. This is a specialized form of specifying the [ABI
+for a function][abi-fn] which enables C to know how to call this function.
+
+[abi-fn]: http://doc.rust-lang.org/reference.html#extern-functions
+
+Finally, if you [take a look at the `Cargo.toml`][cargo-toml] you'll see that
+this library is not compiled as a normal Rust library (rlib) but instead as a
+static archive called a staticlib. This enables all the relevant Rust code to be
```
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: Maybe "which Rust calls a 'staticlib'", since there's nothing special about it.

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+compiled and linked together and the final executable should run and print
+`4 * 2 = 8`.
+
+Rust's lack of a garbage collector and runtime enables this seamless transition
+from C to Rust. The external C code does not need to perform any setup on Rust's
+behalf, making the transition that much cheaper.
+
+### Beyond C
+
+Up to now we've seen how FFI in Rust has 0 overhead and how we can use Rust's
```
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: '0' -> 'zero' I think.

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+[rb2rust]: https://github.com/alexcrichton/rust-ffi-examples/tree/master/ruby-to-rust
+[js2rust]: https://github.com/alexcrichton/rust-ffi-examples/tree/master/node-to-rust
+
+A common desire for writing C code in these languages is to speed up some
+component of a library or application that's performance critical. With the
+features of Rust we've seen here, however, Rust is just as suitable for this
+sort of usage. One of Rust's first production users,
+[Skylight](https://www.skylight.io), is a great example of the benefits that
+Rust can bring even to existing applications by improving the performance and
+memory usage of their data collection agent almost instantly by just using Rust.
```
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: This sentence reads weird to me. I read it as 'Rust improves performance and memory ...' when it is saying 'Skylight improves performance and memory'.

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+
+Many programmers in languages like Python and Ruby have little systems
+programming experience, so reaching down to C is often avoided as it's tough to
+ensure that the program won't crash in a difficult-to-debug way. Rust, however,
```
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: You might avoid ascribing any particular behavior or thinking to large groups of developers. This can probably be phrased in a way that is less accusatory of the deficiencies of Python and Ruby programs - integrating C is just hard and Rust makes it easier.

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+not only brings 0 cost FFI, but *also* the same safety guarantees the the
```
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: '0' -> 'zero' again?

File: _posts/2015-04-24-FFI-and-Rust.md (hunk 11 above)
Comment (alexcrichton) [inline _posts/2015-04-24-FFI-and-Rust.md]: Is there really an extra branch in optimized code though? Any caller of this C function is required to have a check, so there must be at least one. In the case of rust when you `match` the option later on, it's likely to be inlined enough (due to this small implementation) that the conditional for the match and the conditional to create the option to be unified. Which is to say I'm not sure that this actually is imposing extra overhead.

File: _posts/2015-04-24-FFI-and-Rust.md (hunk 16 above)
Comment (alexcrichton) [inline _posts/2015-04-24-FFI-and-Rust.md]: Very good point, I'll shuffle this around a bit.

**Comment by alexcrichton** [comment]
Thanks for the comments everyone! I've pushed some minor edits here and there to address them.

File: _posts/2015-04-24-FFI-and-Rust.md (hunk 11 above)
Comment (brson) [inline _posts/2015-04-24-FFI-and-Rust.md]: I'm not sure either.

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 18]
```diff
@@ -0,0 +1,285 @@
...
+
+Rust's quest for world domination was never destined to happen overnight, so
+Rust needs to be able to interoperate with the existing world just as easily as
+it talks to itself. In particular, **Rust makes it easy to communicate with C
+APIs without overhead, and to leverage its ownership system to provide much
+stronger safety guarantees for those APIs at the same time**.
+
+In more detail, Rust's *foreign function interface* (FFI) is the way that it
+communicated with other languages. Following Rust's design principles, the FFI
+provides a **zero cost abstraction** where function calls between Rust and C
```
Comment (nikomatsakis) [inline _posts/2015-04-24-FFI-and-Rust.md]: Nit: "zero-cost abstraction"

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 19]
```diff
@@ -0,0 +1,285 @@
...
+The `unsafe` block is how the programmer takes responsibility for ensuring
+safety -- you are promising that the actual call you make will not, in fact,
+violate memory safety, and thus that Rust's basic guarantees are upheld.  This
+may seem limiting, but Rust has just the right set of tools to allow consumers
+to not worry about `unsafe` (more on this in a moment).
+
+Now that we've seen how to call a C function from Rust, let's see if we can
+verify this claim of zero overhead. Almost all programming languages can call
+into C one way or another, but it often comes at a cost with runtime type
+conversions or perhaps some language runtime juggling. To get a handle on what
```
Comment (nikomatsakis) [inline _posts/2015-04-24-FFI-and-Rust.md]: Nit: "language-runtime" (language here is an adjective that applies to runtime, not juggling)

File: _posts/2015-04-24-FFI-and-Rust.md [hunk 20]
```diff
@@ -0,0 +1,285 @@
+---
+layout: post
+title: "FFI and Rust"
+author: Alex Crichton
+description: "Zero-cost and safe FFI in Rust"
+---
+
+Rust's quest for world domination was never destined to happen overnight, so
+Rust needs to be able to interoperate with the existing world just as easily as
+it talks to itself. In particular, **Rust makes it easy to communicate with C
```
Comment (nikomatsakis) [inline _posts/2015-04-24-FFI-and-Rust.md]: Maybe s/in particular/for this reason/? I found "in particular" odd here because the structure of the sentence that follows is not parallel to the one before -- that is, this is not a refinement of a more general claim. Before you said "rust needs to be able to do this". I would expect something like, "in particular, rust needs to be able to call C functions" or something, rather than "in particular, rust makes it easy to call C functions".
